### ISSUE TITLE
Save model weights when using dropout layers

### ISSUE THREAD

**Issue by skeydan** [rstudio/keras#633]
Hi Meili,

what I mean is something like this, where you 

- save model weights
- start a new session
- repeat the model definition (but don't train it)
- load the saved weights into the model

Does that work for you?

```
library(keras)
# source("wrapper_concrete_dropout.R")

X_train <- rnorm(100) %>% as.matrix()
Y_train <- rnorm(100) %>% as.matrix()

l <- 1e-4
wd <- l ^ 2 / dim(X_train)[1]
dd <- 2 / dim(X_train)[1]

output_dim <- 1

input <- layer_input(shape = 1)

output <-
  input %>% layer_concrete_dropout(
    layer = layer_dense(units = 32, activation = "relu"),
    weight_regularizer = wd,
    dropout_regularizer = dd
  ) 

mean <-
  output %>% layer_concrete_dropout(
    layer = layer_dense(units = 1),
    weight_regularizer = wd,
    dropout_regularizer = dd
  )

log_var <-
  output %>% layer_concrete_dropout(
    layer_dense(units = 1),
    weight_regularizer = wd,
    dropout_regularizer = dd
  )

output <- layer_concatenate(list(mean, log_var))

model <- keras_model(input, output)

heteroscedastic_loss <- function(y_true, y_pred) {
  mean <- y_pred[, 1:output_dim]
  log_var <- y_pred[, (output_dim + 1):(output_dim * 2)]
  precision <- k_exp(-log_var)
  k_sum(precision * (y_true - mean) ^ 2 + log_var, axis = 2)
}

model %>% compile(
  optimizer = "adam",
  loss = heteroscedastic_loss,
  metrics = c(
    "mse",
    custom_metric("heteroscedastic_loss", heteroscedastic_loss)
  )
)

hist <-
  model %>% fit(X_train,
                Y_train,
                epochs = 1,
                batch_size = batch_size,
                verbose = 0)

# save model weights only
model %>% save_model_weights_hdf5("/tmp/m.hdf5")


##################################################################
### restart R and define the model again, but don't train it
##################################################################

library(keras)
# source("wrapper_concrete_dropout.R")

X_train <- rnorm(100) %>% as.matrix()
Y_train <- rnorm(100) %>% as.matrix()

l <- 1e-4
wd <- l ^ 2 / dim(X_train)[1]
dd <- 2 / dim(X_train)[1]

output_dim <- 1

input <- layer_input(shape = 1)

output <-
  input %>% layer_concrete_dropout(
    layer = layer_dense(units = 32, activation = "relu"),
    weight_regularizer = wd,
    dropout_regularizer = dd
  ) 

mean <-
  output %>% layer_concrete_dropout(
    layer = layer_dense(units = 1),
    weight_regularizer = wd,
    dropout_regularizer = dd
  )

log_var <-
  output %>% layer_concrete_dropout(
    layer_dense(units = 1),
    weight_regularizer = wd,
    dropout_regularizer = dd
  )

output <- layer_concatenate(list(mean, log_var))

model <- keras_model(input, output)

heteroscedastic_loss <- function(y_true, y_pred) {
  mean <- y_pred[, 1:output_dim]
  log_var <- y_pred[, (output_dim + 1):(output_dim * 2)]
  precision <- k_exp(-log_var)
  k_sum(precision * (y_true - mean) ^ 2 + log_var, axis = 2)
}

model %>% compile(
  optimizer = "adam",
  loss = heteroscedastic_loss,
  metrics = c(
    "mse",
    custom_metric("heteroscedastic_loss", heteroscedastic_loss)
  )
)

# load saved weights
model %>% load_model_weights_hdf5("/tmp/m.hdf5")

```

**Comment by meilibaragatti** [comment]
Thank you Skeydan, this is working for me now!
Meili

**Comment by skeydan** [comment]
Great, I'm glad to hear that! Let me know if you are able to successfully use this in a study (we would be happy to have a post on the TF blog :-))
